### PR TITLE
Minimized PR: Make the wellModel self-contained

### PR DESCRIPTION
**Pull request by totto82**
This PR contain the main part of PR #1313 

The wellModel is now persistent over the time steps,
with an update method called every reportStep/episode.

This allows the following simplifications:

    1. move the wellState to the WellModel
    2. add a ref to the ebosSimulator to the wellModel
    3. clean up the parameters passed to the wellModel methods
    4. move RESV handling to the WellModel and the rateConverter
    5. move the econLimit update to the WellModel

### CODE REVIEW

**Comment by totto82** [comment]
To make sure I didn't messed up something during the commit split I will rerun it through my local tests.

**Comment by totto82** [comment]
jenkins build this please

File: opm/autodiff/BlackoilWellModel.hpp [hunk 1]
```diff
@@ -126,44 +115,59 @@ namespace Opm {
...
 
-            int numWells() const;
+            // Check if well equations is converged.
+            bool getWellConvergence(const std::vector<Scalar>& B_avg) const;
 
-            /// return true if wells are available in the reservoir
-            bool wellsActive() const;
+            // return all the wells.
+            const WellCollection& wellCollection() const;
+            WellCollection& wellCollection();
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel.hpp]: This non-const method should not need to be public.

File: opm/autodiff/BlackoilWellModel.hpp [hunk 2]
```diff
@@ -126,44 +115,59 @@ namespace Opm {
...
 
-            void setWellsActive(const bool wells_active);
+            // return the internal well state, ignore the passed one.
+            // Used by the legacy code to make it compatible with the legacy well models.
+            const WellState& wellState(const WellState& well_state OPM_UNUSED) const { return wellState(); }
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel.hpp]: I'd prefer to see all method implementations (even small ones) in the _impl.hpp file, since that makes it much easier to see where a member is used etc. (need only search one file). Taking that argument further, one can argue in favour of having it all in the hpp file, but for big classes like this it is important to have a readable public interface, so I think we should keep the two files.

File: opm/autodiff/BlackoilWellModel.hpp [hunk 3]
```diff
@@ -126,44 +115,59 @@ namespace Opm {
...
+            void setRestartWellState(const WellState& well_state) { previous_well_state_ = well_state; }
 
-            /// upate the dynamic lists related to economic limits
-            void updateListEconLimited(const Schedule& schedule,
-                                       const int current_step,
-                                       const Wells* wells_struct,
-                                       const WellState& well_state,
-                                       DynamicListEconLimited& list_econ_limited) const;
+            // called at the beginning of a time step
+            void beginTimeStep() {
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel.hpp]: Implementation to _impl.hpp, as above (and others below).

File: opm/autodiff/BlackoilWellModel.hpp [hunk 4]
```diff
@@ -126,44 +115,59 @@ namespace Opm {
...
-            bool wells_active_;
-            const Wells*   wells_;
-            const std::vector< const Well* > wells_ecl_;
+            // called at the beginning of a report step
+            void beginReportStep(const int time_step);
+
+            // called at the end of a report step
+            void endReportStep() {
+                // update the list contanining information of closed wells
+                // and connections due to ecnomical limits
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel.hpp]: Typo: ecnomical

File: opm/autodiff/BlackoilWellModel.hpp [hunk 5]
```diff
@@ -175,59 +179,60 @@ namespace Opm {
...
+            SimulatorReport last_report_;
 
             // used to better efficiency of calcuation
             mutable BVector scaleAddRes_;
 
-            void updateWellControls(WellState& xw) const;
+            const Wells* wells() const { return wells_manager_->c_wells(); }
+
+            const Schedule& schedule() const
+            { return ebosSimulator_.gridManager().schedule(); }
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel.hpp]: Whils this line is ok for now, it is likely that we will change it in the near future, to support external (to the ebos simulator and well model classes) modification of schedules.

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 6]
```diff
@@ -5,139 +5,168 @@ namespace Opm {
...
+                                                schedule(),
+                                                timeStepIdx,
+                                                Opm::UgGridHelpers::numCells(grid),
+                                                Opm::UgGridHelpers::globalCell(grid),
+                                                Opm::UgGridHelpers::cartDims(grid),
+                                                Opm::UgGridHelpers::dimensions(grid),
+                                                Opm::UgGridHelpers::cell2Faces(grid),
+                                                Opm::UgGridHelpers::beginFaceCentroids(grid),
+                                                dynamic_list_econ_limited_,
+                                                grid.comm().size()>1,
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: Space around binary operators.

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 7]
```diff
@@ -5,139 +5,168 @@ namespace Opm {
...
+                                                defunct_well_names) );
+	
+        // Wells are active if they are active wells on at least
+        // one process.
+        wells_active_ = localWellsActive() ? 1 : 0;
+        wells_active_ = grid.comm().max(wells_active_);
+
+        // The well state initialize bhp with the cell pressure in the top cell.
+        // We must therefore provide it with updated cell pressures
+        size_t nc = Opm::UgGridHelpers::numCells(grid);
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: You already have number_of_cells_ (for local partition).

File: opm/autodiff/BlackoilWellModel.hpp (hunk 5 above)
Comment (andlaus) [reply opm/autodiff/BlackoilWellModel.hpp]: as far as I can see, that line should be okay even after that because I'm pretty sure that you will want to update the gridManager's schedule in that case?! (In other words: what's the point of a schedule object that the simulation knows nothing about?)

File: opm/autodiff/BlackoilWellModel.hpp (hunk 1 above)
Comment (totto82) [reply opm/autodiff/BlackoilWellModel.hpp]: It seems that the call to 
wellModel().wellCollection().groupControlActive()
and
wellModel().wellCollection().groupTargetConverged(wellModel().wellState().wellRates()
in #250 in BlackoilModelEbos whant's non const wellCollections.

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 8]
```diff
@@ -5,139 +5,168 @@ namespace Opm {
...
-    }
-
-
 
+        // compute VFP properties
+        vfp_properties_.reset (new VFPProperties (
+                                   eclState.getTableManager().getVFPInjTables(),
+                                   eclState.getTableManager().getVFPProdTables()) );
 
+        // update the previous well state. This is used to restart failed restarts.
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: Should probably be "... to restart failed steps"?

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 9]
```diff
@@ -5,139 +5,168 @@ namespace Opm {
...
-
     template<typename TypeTag>
     std::vector<typename BlackoilWellModel<TypeTag>::WellInterfacePtr >
     BlackoilWellModel<TypeTag>::
-    createWellContainer(const Wells* wells,
-                        const std::vector< const Well* >& wells_ecl,
-                        const bool use_multisegment_well,
-                        const int time_step,
-                        const ModelParameters& param)
+    createWellContainer(const int time_step)
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: I think this should be const.

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 10]
```diff
@@ -164,38 +193,42 @@ namespace Opm {
...
 
         if (iterationIdx == 0) {
-            prepareTimeStep(ebosSimulator, well_state);
+            prepareTimeStep();
+        }
+
+
+
+        if ( ! wellsActive() ) {
+            return;
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: Why is this moved down? It would be better to have it before the calls to updatePerforationIntensiveQuantities() and prepareTimeStep().

File: opm/autodiff/BlackoilWellModel.hpp (hunk 1 above)
Comment (atgeirr) [reply opm/autodiff/BlackoilWellModel.hpp]: That is very strange, as both those methods of WellCollection are const! There must be something else preventing this. Perhaps there is no const overload of wellModel() in the BlackoilModelEbos class?

File: opm/autodiff/BlackoilWellModel.hpp (hunk 5 above)
Comment (atgeirr) [reply opm/autodiff/BlackoilWellModel.hpp]: Quite possibly true.

**Comment by atgeirr** [comment]
We have now read through the PR (we have not checked that moved functions are identical, but trust that it is done correctly as tests indicate that it's ok). We do not have any more comments on details in the PR, our remaining comments apply to the PR as a whole.

Your list of changes is slightly incomplete:

> 1. move the wellState to the WellModel
> 2. add a ref to the ebosSimulator to the wellModel
> 3. clean up the parameters passed to the wellModel methods
> 4. move RESV handling to the WellModel and the rateConverter
> 5. move the econLimit update to the WellModel

The missing "item 0" would be: move the well model to outside the time loop and update it instead of recreating.

We think (starting from the bottom) that 4 and 5 are great changes. We are not quite comfortable with 2 and 3, since we fear it makes the coupling between the ebos simulator and the well model stronger, but refactoring that is a separate issue and we accept that given the current state it does make the code simpler.

Item 1 (and 0) we are not happy with, for the following reasons:
 - We think that the state objects had an important role to play; separating state from model/behaviour is a good thing.
 - Treatment of state and updating of it is now less transparent.
 - It is no longer feasible to manipulate state (well or reservoir) in the Simulator::run() method or after.
 - Now the BlackoilModelEbos is created for each report step, while the well model is created only once and then updated. That is a little less obvious and more messy than creating both. In particular since the well model is the one with least continuity across report steps. The well state (and its ability to copy partial state from a previous well state object) is sufficient to takes care of the continuity, if any.

Still, we will accept this. But we feel that a thorough discussion of how we treat (and should treat) the state variables (both reservoir and well) is called for. Perhaps at the next hangout?

I'll merge this when all minor comments are addressed, but first I'd like to verify that it does not change anything for model 2 with multisegment wells turned on (we are running it this afternoon, so I'll probably merge tomorrow).

**Comment by totto82** [comment]
Thanks for the review.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by andlaus** [comment]
> Now the BlackoilModelEbos is created for each report step, while the well model is created only once and then updated. That is a little less obvious and more messy than creating both. In particular since the well model is the one with least continuity across report steps. The well state (and its ability to copy partial state from a previous well state object) is sufficient to takes care of the continuity, if any.

I agree that the lifetime change makes things a bit messy, but IMO the `BlackoilModelEbos` objects should match the lifetime assumptions of eWoms' simulator/model/problem objects in the not-too-distant future, i.e.  like the `ebosSimulator_` attribute, they should be persistent for the whole simulation. you should see this PR as a first step into that direction.

(The well model should be seen as the "well manager" in the sense that it deals with all well-related things, not just with the specific configuration at a given point in time.)

**Comment by GitPaean** [comment]
A full running of model 2 with activating multisegment wells did not show any change of simulation behavior  in anyway due to this PR.

**Comment by GitPaean** [comment]
> (The well model should be seen as the "well manager" in the sense that it deals with all well-related things, not just with the specific configuration at a given point in time.)

Here are my two cents. 
`Simulator` class as the main drive of the simulation should be persistent during the whole simulation. `State`s are used to record or recover a running, are the ones should be passed between time/report steps. `State`s are the input and output (updated by) of `Model`s and `Solver`s. Between report steps, there might be some, while there might be no continuity at all (especially for the well models) from Schedule. Smaller while decoupled classes can be easier to maintain and extend.

**Comment by atgeirr** [comment]
> A full running of model 2 with activating multisegment wells did not show any change of simulation behavior in anyway due to this PR.

I am satisfied testing is ok, then.

Thanks for the work, while we may not agree about everything it most certainly simplifies certain things and move them in the right direction.

**Comment by totto82** [comment]
Norne, Model  2 rel 0 and rel 5 also shows no change.
